### ISSUE TITLE
Problems inferring types of lambda functions

### ISSUE THREAD

**Issue by RaulGracia** [javaparser/javasymbolsolver#186]
Hi all,

I recently started to work with JavaSymbolSolver. My objective is to extract the method calls of an Stream variable (map, flatMap, filter,...) in a piece of code and then infer the type of the lambda function passed as parameter. 

For instance, consider the following dummy code:

```
List<String> myList = Arrays.asList("a1", "a2", "b1", "c2", "c1");

Stream<String> myStream = myList.stream();

myStream.map(s -> s + ":)")
	        .filter(s -> s.length()>1)
		.flatMap(s -> Arrays.stream(s.split("")))
		.map(s -> new SimpleEntry<String, Integer>(":)",1))
		.map(x -> x.toString())
		.filter(x -> x.length() > 10);
```

The idea is to write a program that combines JavaParser and JavaSymbolSolver to iterate over the AST graph and then infer the types of input lambdas. For example, something like:
```
CompilationUnit cu = JavaParser.parse(in);
Navigator.findAllNodesOfGivenClass(cu, LambdaExpr.class).stream()
        .forEach(l` -> System.out.println(javaParserFacade.getType(l, true).describe()));
```

The output of the above code for the previous dummy example on a Stream object outputs the following:

> java.util.function.Function<? super T, ? extends java.lang.String>
> java.util.function.Predicate<? super java.lang.String>
> Exception in thread "main" java.lang.IllegalStateException: Equivalent types are: [ReferenceType{java.util.stream.Stream,...

As can be observed, for "map" and "filter" JavaSymbolSolver can achieve type inference. In the case where it fails, my code tries to infer the types of lambdas by inspecting the previous/next lambdas (which is better, but not a silver bullet).

However, the most problematic part comes when JavaSymbolSolver tries to infer the type of a lambda passed to a flatMap method call, which actually fails throwing an IllegalStateException. My question is: is this exception reflecting a limitation of JavaSymbolSolver or could be due to a bad use from my side? Is JavaSymbolSolver a suitable tool to infer the types of lambdas in code as the first example I described?

Thanks a lot,
Raúl.

PS: I posted this issue by request of Federico Tomassetti.

**Comment by xdrop** [comment]
This is a bug in JSS. What seems to happen is:

It first tries to infer type variables using argument types. 

It sees:
Formal type:
`Function<T,R>`

Actual type:
`Function<? super String, ? extends Stream<R>>`

So the inference context establishes equivalence `T=> ? super String` and `R => ? extends Stream<R>`

---

It then carries on to establish any equivalences based on the return type

And also establishes an equivalence between `R => Stream<String>`.

-- 

When it tries to resolve there are two equivalent types to `R` and it doesn't know which one to choose throwing the exception

**Comment by ftomassetti** [comment]
@RaulGracia can you give it a try using the version currently on master?
@xdrop has just provided a solution for the specific code you shared, let us know if you are running in more issues

**Comment by RaulGracia** [comment]
Hi all,

As suggested by @ftomassetti , I tried the fix provided by @xdrop in the master branch. I have to say the current version significantly improves the results of the previous one, so congratulations! :)

In fact, I did a small evaluation of JSS with the with and without the new fix. I tried 4 simple examples that I'm using during the initial development phase of the software that uses JSS (the results are appended at the end of the text).

While in most cases the fix provides better results than the previous version, I detected a case in which the new version fails whereas the old one correctly infers the types of lambdas (SimpleLogAnalyzer2.java). In particular, the new version of JSS does not detect that the return argument of the Function map(l -> l.length()) returns an Integer. As can be observed in the results, the old version of the code correctly infers the return type of this function, despite in the subsequent Predicate (filter(s -> s > 15)) the type detected is the primitive type (int) instead of the class (Integer).

As the previous code already did the job for that case, I hope that it might not be hard to merge with the new code to avoid the regression.

My work in the next weeks (or months) will be related to the inspection of code, so I will exercise JSS in many scenarios, specifically for lambdas. I also hope that my feedback could help to improve the code :)

Thanks a lot,
Raúl.

**RESULTS OF OLD/NEW JSS FOR 4 SIMPLE SETS OF LAMBDAS**

> HelloWorld.java

```
List<String> myList = Arrays.asList("a1", "a2", "b1", "c2", "c1");
Stream<String> myStream = myList.stream();
myStream.map(s -> s + ":)")
	.filter(s -> s.length() > 1)
	.flatMap(s -> Arrays.stream(s.split("")))
	.map(s -> new SimpleEntry<String, Integer>(":)", 1))
	.map(x -> x.toString())
	.filter(x -> x.length() > 10);
```

_OLD JSS_

(**OK**) Type found by JSS: java.util.function.Function<T, java.lang.String>
(**OK**) Type found by JSS: java.util.function.Predicate<java.lang.String>
(**FAILURE**) Unable to find type for lambda: s -> Arrays.stream(s.split(""))
(**WRONG INFERENCE**) Type found by JSS: java.util.function.Function<java.util.AbstractMap.SimpleEntry<java.lang.String, java.lang.Integer>, java.util.AbstractMap.SimpleEntry<java.lang.String, java.lang.Integer>>
(**OK**) Type found by JSS: java.util.function.Function<java.util.AbstractMap.SimpleEntry<java.lang.String, java.lang.Integer>, java.lang.String>
(**FAILURE**) Unable to find type for lambda: x -> x.length() > 10

_NEW JSS_ 

(**OK**) Type found by JSS: java.util.function.Function<T, java.lang.String>
(**OK**) Type found by JSS: java.util.function.Predicate<java.lang.String>
(**OK**) Type found by JSS: java.util.function.Function<java.lang.String, java.util.stream.Stream<java.lang.String>>
(**OK**) Type found by JSS: java.util.function.Function<java.lang.String, java.util.AbstractMap.SimpleEntry<java.lang.String, java.lang.Integer>>
(**OK**) Type found by JSS: java.util.function.Function<java.util.AbstractMap.SimpleEntry<java.lang.String, java.lang.Integer>, java.lang.String>
(**OK**) Type found by JSS: java.util.function.Predicate<java.lang.String>

> SimpleLogAnalyzer.java

```
Stream<String> myStream = Files.lines(Paths.get(inputFile));
myStream.filter(s -> s.contains("Hamlet"))
	.map(l -> l.toString())
	.filter(s -> s.length() > 10)
	.count()
```

_OLD JSS_

(**WRONG INFERENCE**) Type found by JSS: java.util.function.Predicate<T>
(**OK**) Type found by JSS: java.util.function.Function<java.lang.String, java.lang.String>
(**OK**) Type found by JSS: java.util.function.Predicate<java.lang.String>

_NEW JSS_

(**WRONG INFERENCE**) Type found by JSS: java.util.function.Predicate<T>
(**OK**) Type found by JSS: java.util.function.Function<java.lang.String, java.lang.String>
(**OK**) Type found by JSS: java.util.function.Predicate<java.lang.String>

> SimpleLogAnalyzer2.java

```
Stream<String> myStream = Files.lines(Paths.get(inputFile));
myStream.filter(s -> s.contains("Hamlet"))
	.map(l -> l.length())
	.filter(s -> s > 15)
	.count()
```

_OLD JSS_

(**OK**) Type found by JSS: java.util.function.Predicate<T>
(**OK**) Type found by JSS: java.util.function.Function<java.lang.String, java.lang.Integer>
(**OK**) Type found by JSS: java.util.function.Predicate<int>

_NEW JSS_

(**WRONG INFERENCE**) Type found by JSS: java.util.function.Predicate<T>
(**WRONG INFERENCE**) Type found by JSS: java.util.function.Function<java.lang.String, R>
(**WRONG INFERENCE**) Type found by JSS: java.util.function.Predicate<T>


> WordCountJava8Streams.java

```
Stream<String> myStream = Files.lines(Paths.get(inputFile));
myStream.flatMap(line -> Arrays.stream(line.trim().split(" ")))
	.map(word -> word.replaceAll("[^a-zA-Z]", "").toLowerCase().trim())
	.map(word -> new SimpleEntry<String, Long>(word, (long) 1))
	.collect(groupingBy(SimpleEntry::getKey, counting()))
```

_OLD JSS_

(**FAILURE**) Unable to find type for lambda: word -> new SimpleEntry<String, Long>(word, (long) 1)
(**FAILURE**) Unable to find type for lambda: word -> word.replaceAll("[^a-zA-Z]", "").toLowerCase().trim()
(**FAILURE**) Unable to find type for lambda: line -> Arrays.stream(line.trim().split(" "))

_NEW JSS_

(**OK**) Type found by JSS: java.util.function.Function<T, java.util.stream.Stream<java.lang.String>>
(**OK**) Type found by JSS: java.util.function.Function<java.lang.String, java.lang.String>
(**OK**) Type found by JSS: java.util.function.Function<java.lang.String, java.util.AbstractMap.SimpleEntry<java.lang.String, java.lang.Long>>

**Comment by ftomassetti** [comment]
@RaulGracia thank you so much, this feedback is very precious!
Thanks to @xdrop for the huge improvement.
Getting this stuff to work 100% is hard but precise reports like this one help us moving in the right direction

**Comment by xdrop** [comment]
Thanks @RaulGracia! I'd highly appreciate once #188 gets merged to get another report with what works and what not.

**Comment by ftomassetti** [comment]
It is now merged, looking forward to @RaulGracia feedback

**Comment by RaulGracia** [comment]
Hi @ftomassetti and @xdrop,

I tested the new fix against the same 4 programs, and I obtained very good results :)

The new version of JSS correctly infers the types for SimpleLogAnalyzer2.java lambas, and the rest of sample programs are also correctly analyzed. And not only this, I also realized that now JSS also provides type inference for the first lambda executed on a Stream, which is something that was apparently not working before. In summary, this means that I haven't seen any Predicate<R> or Function<T, String> in any of the programs tested, which is a great improvement compared to the prior version.

I think that this issue can be happily closed if you want; anyway, I will continue working with JSS and lambdas, so I will provide you with more feedback if I detect any other problem for some specific case.

Thanks a lot!
Raúl.
